### ISSUE TITLE
Implementing a custom torch component and a preprocessing layer to handle Docs

### ISSUE THREAD

**Issue by IsmailShaheen** [explosion/spaCy#11313]
I am trying to implement a simple token classifier with pytorch and encapsulate it in a custom trainable component. I know that the model the component will use should take a list of Docs as an input. I tried to implement a simple thinc model that takes the transformer output stored in `doc._.trf_data` from the doc list and passes it to the torch model as a 2d tensor input and chaining the two together. I am attaching the code sample I am using for clarification. I am planning to use a transformer as a feature extractor while using the classifier. I also tried loading the `en_core_web_trf` pipeline as a side experiment to check the `token.tensor` and `token.vector` fields but they both do not contain any values. That's why I decided to access `doc._.trf_data` directly.

My question is: If I am to implement the preprocessor layer as follows, how can I implement the backprop callback to store the gradient in the doc object to be passed back to the transformer for training?
I am also open to suggestions if I am approaching this in the wrong way. All I need to do for now is get the transformer output and pass it to a pytorch model that takes 2d tensors as input and not a list of doc objects

``` python
@spacy.registry.architectures("LinearPreprocessor.v1")
def create_linear_preprocessor():

    def preprocessor_forward(
        model: Model[List[Doc], Floats2d],
        docs: List[Doc],
        is_train: bool,
    ) -> Tuple[Floats2d, Callable]:
        Y = []
        for doc in docs:
            embed_size = doc._.trf_data.model_output.last_hidden_state.shape[-1]
            Y.append(doc._.trf_data.model_output.last_hidden_state.reshape(-1, embed_size))
        Y = model.ops.asarray2f(np.concatenate(Y, axis=0))

        def backprop(dY):
            return docs #TODO: implement backprop

        return Y, backprop

    preprocessor_model = Model(
        "preprocessor",                     # string name of layer
        preprocessor_forward,               # forward function
    )
    return preprocessor_model


@spacy.registry.architectures("TorchFeedForward.v1")
def create_torch_feedforward(
    nO: int,            # lables_number
    width: int,         # hidden_size
    hidden_width: int,
    dropout: float
) -> Model[Floats2d, Floats2d]:
    torch_model = nn.Sequential(
        nn.Linear(width, hidden_width),
        nn.Dropout(dropout),
        nn.ReLU(),
        nn.Linear(hidden_width, nO),
        nn.Softmax(dim=-1) if nO > 1 else nn.Sigmoid()
    )  # (batch_size x sequence_length, hidden_size) -> (batch_size x sequence_length, lables_number)
    return PyTorchWrapper(torch_model)


@spacy.registry.architectures("TokenClassifierModel.v1")
def create_model(
    preprocessor: Model[List[Doc], Floats2d],
    classifier: Model[Floats2d, Floats2d],
) -> Model[List[Doc], Floats2d]:
    model = chain(preprocessor, with_array(classifier))
    return model
```

**Comment by polm** [comment]
Since this is an implementation question rather than a bug report, I'll move it to Discussions.
